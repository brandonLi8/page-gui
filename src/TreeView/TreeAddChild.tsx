import React, { Component, ChangeEvent, FormEvent, Fragment } from 'react';
import { ComponentStructure } from '../types';
import Select from 'react-select';

export interface TreeAddChildProps {
  onAddComponent: (componentName: string) => any;
  componentStructures: ComponentStructure[];
}

export interface TreeAddChildState {
  selected: null | { label: string, value: string };
}

export default class TreeAddChild extends Component<TreeAddChildProps, TreeAddChildState> {
  state: TreeAddChildState = {
    selected: null
  }

  handleAdd = () => {
    const { selected } = this.state;
    if (!!selected) {
      const componentInfo = this.props.componentStructures.find(value => value.name === selected.value);
      if (!!componentInfo) {
        this.props.onAddComponent(selected.value);
        this.setState({ selected: null });
      }
    }
  }

  handleChange = (selected) => {
    this.setState({ selected });
  }

  public render() {
    const options = this.props.componentStructures.map(info => ({
      value: info.name,
      label: info.name
    }));
    return (
      <Fragment>
        <Select 
          value={this.state.selected}
          onChange={this.handleChange}
          options={options}
        />
        <button onClick={this.handleAdd}>Add Component</button>
      </Fragment>
    );
  }
}