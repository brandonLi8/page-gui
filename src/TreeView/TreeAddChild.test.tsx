import React, { Component, Fragment } from 'react';
import { mount } from 'enzyme';
import Select from 'react-select';
import TreeAddChild from './TreeAddChild';
import { ComponentStructure } from '../types';

class ExampleComponent extends Component {
  render() {
    return <Fragment>Hello</Fragment>;
  }
}

describe('TreeAddChild', () => {
  let addComponent, componentStructures, submitEvent;
  beforeEach(() => {
    addComponent = jest.fn();
    componentStructures = [
      {
        component: ExampleComponent,
        name: 'First',
        propertyTypes: {}
      },
      {
        component: ExampleComponent,
        name: 'Second',
        propertyTypes: {}
      }
    ] as ComponentStructure[];
    submitEvent = {
      preventDefault: jest.fn()
    }
  });
  it('defaults to nothing', () => {

  });
  it('displays a list of components', () => {
    const wrapper = mount(<TreeAddChild onAddComponent={addComponent} componentStructures={componentStructures} />);
    expect(wrapper.find(Select).first().props().options).toMatchSnapshot();
  });
  describe('when Add button is clicked', () => {
    let wrapper;
    beforeEach(() => {
      wrapper = mount(<TreeAddChild onAddComponent={addComponent} componentStructures={componentStructures} />);
      // Select the second option
      wrapper.find(Select).first().props().onChange({ value: 'Second', label: 'Second' });
      // Click
      wrapper.find('button').last().simulate('click');
    });
    it('calls onAddComponent', () => {
      expect(addComponent).toBeCalledWith('Second');
    });
    it('remove selected component value', () => {
      expect(wrapper.find(Select).last().props().value).toBe(null);
    })
  })
});