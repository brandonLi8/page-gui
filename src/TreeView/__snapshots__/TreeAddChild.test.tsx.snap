// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeAddChild displays a list of components 1`] = `
Array [
  Object {
    "label": "First",
    "value": "First",
  },
  Object {
    "label": "Second",
    "value": "Second",
  },
]
`;
